File: src/app_web.zig
const app = @import("app.zig");
const main = @import("main.zig");
const std = @import("std");

const js = struct {
    const CanvasId = u32;
    const DomHighResTimeStamp = f64;

    extern fn setWindowTitle(wasm_id: main.WasmId, title_ptr: [*]const u8, title_len: usize) void;
    extern fn createCanvas(wasm_id: main.WasmId, width: u32, height: u32) CanvasId;
    extern fn destroyCanvas(canvas_id: CanvasId) void;
    extern fn now() DomHighResTimeStamp;
};

pub const Window = struct {
    id: js.CanvasId,
    width: u32,
    height: u32,

    pub fn init(desc: app.WindowDesc) !Window {
        if (desc.title.len > 0) {
            js.setWindowTitle(main.wasm_id, desc.title.ptr, desc.title.len);
        }
        return Window{
            .id = js.createCanvas(main.wasm_id, desc.width, desc.height),
            .width = desc.width,
            .height = desc.height,
        };
    }

    pub fn deinit(window: *Window) void {
        const empty: []const u8 = &.{};
        js.setWindowTitle(main.wasm_id, empty.ptr, empty.len);
        js.destroyCanvas(window.id);
    }

    pub fn getWidth(window: Window) u32 {
        return window.width;
    }

    pub fn getHeight(window: Window) u32 {
        return window.height;
    }
};

pub const Timer = struct {
    start_time: js.DomHighResTimeStamp,

    pub fn start() !Timer {
        return Timer{ .start_time = js.now() };
    }

    pub fn read(self: Timer) u64 {
        return timeStampToNs(js.now() - self.start_time);
    }

    pub fn reset(self: *Timer) void {
        self.start_time = js.now();
    }

    pub fn lap(self: *Timer) u64 {
        var now = js.now();
        var lap_time = self.timeStampToNs(now - self.start_time);
        self.start_time = now;
        return lap_time;
    }

    fn timeStampToNs(duration: js.DomHighResTimeStamp) u64 {
        return @floatToInt(u64, duration * 1000000.0);
    }
};

File: src/app_web.js
const _BaseWindowTitle = document.title;

const _app = {
    _canvases: new _Objs(),
    _canvasParent: document.body,
    _observer: new IntersectionObserver(function(_entries) {
        _entries.forEach(_entry => {
            for (let i = _app._canvases._begin();
                i < _app._canvases._end();
                i = _app._canvases._next(i))
            {
                const _canvas = _app._canvases._get(i);
                if (_canvas._obj === _entry.target) {
                    _main._wasms._get(_canvas._wasmId)._canUpdate = _entry.isIntersecting;
                }
            }
        });
    }),

    setWindowTitle(_wasmId, _titlePtr, _titleLen) {
        document.title = _titleLen > 0 ?
            _main._getString(_wasmId, _titlePtr, _titleLen) :
            _BaseWindowTitle;
    },

    createCanvas(_wasmId, _width, _height) {
        const _canvas = document.createElement("canvas");
        _canvas.width = _width;
        _canvas.height = _height;
        _app._canvasParent.appendChild(_canvas);
        _app._observer.observe(_canvas);
        return _app._canvases._insert({
            _obj: _canvas,
            _wasmId: _wasmId,
            _parent: _app._canvasParent
        });
    },
    
    destroyCanvas(_canvasId) {
        const _canvas = _app._canvases._get(_canvasId);
        _canvas._parent.removeChild(_canvas._obj);
        _app._canvases._remove(_canvasId);
    },

    _setCanvasParent(_canvasParent) {
        _app._canvasParent = _canvasParent === null ?
            document.body :
            document.getElementById(_canvasParent);
    },

    now() {
        return performance.now();
    },
};
